ompi/mca/osc/rdma/osc_rdma_passive_target.c

```diff
@@ -309,6 +442,53 @@ int ompi_osc_rdma_unlock_atomic (int target, ompi_win_t *win)
     return ret;
 }
 
+int ompi_osc_rdma_iunlock_atomic (int target, ompi_win_t *win)
+{
+    ompi_osc_rdma_module_t *module = GET_MODULE(win);
+    ompi_osc_rdma_peer_t *peer;
+    ompi_osc_rdma_sync_t *lock;
+    int ret = OMPI_SUCCESS;
+
+    OPAL_THREAD_LOCK(&module->lock);
+
+    OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_TRACE, "iunlock: %d, %s", target, win->w_name);
+
+    lock = ompi_osc_rdma_module_lock_find (module, target, &peer);
+    if (OPAL_UNLIKELY(NULL == lock)) {
+        OSC_RDMA_VERBOSE(MCA_BASE_VERBOSE_INFO, "target %d is not locked in window %s",
+                         target, win->w_name);
+        OPAL_THREAD_UNLOCK(&module->lock);
+        return OMPI_ERR_RMA_SYNC;
+    }
+
+    ompi_osc_rdma_module_lock_remove (module, lock);
+
+    /* finish all outstanding fragments */
+    if(ompi_osc_rdma_sync_rdma_icomplete (lock) == false){
+        return OMPI_ERR_RMA_SYNC;
```

Comment:
Must release thread lock before exiting; needs this line inside `if`, before `return`:
```OPAL_THREAD_UNLOCK(&module->lock);```

Reply:
This might not be the right response to `icomplete == false`

In the blocking version, this would block until all fragments are complete. For the nonblocking version, we probably want to return with "not done yet", i.e. not a permanent failure but a value that means "try again in a little while".

I'd expect this to be called in a loop - or from the progress engine, which will call it repeatedly (i.e. same effect as a loop).

```unlock()```
is the same as:
```while (false == iunlock());```
but the latter can be broken up and done over many different MPI calls, with each invocation of MPI progress doing:
```if (false == inunlock()) {return_without_changing_request_status;} else {change_status_of_request;}```

We probably need a new error code; something like `OMPI_ERR_RMA_NB_PENDING`.

Reply:
Inside ```ompi_osc_rdma_try_unlock_atomic()``` we will have
```
if(OMPI_ERR_RMA_NB_PENDING == ompi_osc_rdma_sync_rdma_icomplete (lock) ){
      OPAL_THREAD_UNLOCK(&module->lock);
      return OMPI_ERR_RMA_NB_PENDING;
}
```
Whoever is calling ```ompi_osc_rdma_try_unlock_atomic()``` must call it in a loop until it gets ```OMPI_SUCCESS```. Is this correct?

Reply:
Yes, that's right.